core/expr.rs

```diff
@@ -0,0 +1,583 @@
+
+use crate::{
+    function::{Func, SingleRowFunc},
+    schema::{Column, Schema, Table},
+    select::{ColumnInfo, Select, SrcTable},
+    util::normalize_ident,
+    vdbe::{BranchOffset, Insn, ProgramBuilder},
+};
+
+pub fn build_select<'a>(schema: &Schema, select: &'a ast::Select) -> Result<Select<'a>> {
+    match &select.body.select {
+        ast::OneSelect::Select {
+            columns,
+            from: Some(from),
+            where_clause,
+            ..
+        } => {
+            let (table_name, maybe_alias) = match &from.select {
+                Some(select_table) => match select_table.as_ref() {
+                    ast::SelectTable::Table(name, alias, ..) => (
+                        &name.name,
+                        alias.as_ref().map(|als| match als {
+                            ast::As::As(alias) => alias,     // users as u
+                            ast::As::Elided(alias) => alias, // users u
+                        }),
+                    ),
+                    _ => todo!(),
+                },
+                None => todo!(),
+            };
+            let table_name = &table_name.0;
+            let maybe_alias = maybe_alias.map(|als| &als.0);
+            let table = match schema.get_table(table_name) {
+                Some(table) => table,
+                None => anyhow::bail!("Parse error: no such table: {}", table_name),
+            };
+            let mut joins = Vec::new();
+            joins.push(SrcTable {
+                table: Table::BTree(table.clone()),
+                alias: maybe_alias,
+                join_info: None,
+            });
+            if let Some(selected_joins) = &from.joins {
+                for join in selected_joins {
+                    let (table_name, maybe_alias) = match &join.table {
+                        ast::SelectTable::Table(name, alias, ..) => (
+                            &name.name,
+                            alias.as_ref().map(|als| match als {
+                                ast::As::As(alias) => alias,     // users as u
+                                ast::As::Elided(alias) => alias, // users u
+                            }),
+                        ),
+                        _ => todo!(),
+                    };
+                    let table_name = &table_name.0;
+                    let maybe_alias = maybe_alias.as_ref().map(|als| &als.0);
+                    let table = match schema.get_table(table_name) {
+                        Some(table) => table,
+                        None => anyhow::bail!("Parse error: no such table: {}", table_name),
+                    };
+                    joins.push(SrcTable {
+                        table: Table::BTree(table),
+                        alias: maybe_alias,
+                        join_info: Some(join),
+                    });
+                }
+            }
+
+            let _table = Table::BTree(table);
+            let column_info = analyze_columns(columns, &joins);
```

Comment:
I would have put `build_select()` into `select.rs`, but `expr.rs´ depends on `select.rs` and `analyze_columns()` would make `select.rs` circularly depend on `expr.rs` -- `analyze_columns() calls `analyze_column()` which in turn calls `analyze_expr()`

I'm sure these will be shuffled around many times